File: app/src/main/cpp/wn-steam-client/src/encrypted_channel.cpp
#include "wn_steam/encrypted_channel.h"

#include <android/log.h>

#include <algorithm>
#include <cstring>
#include <exception>

#include "wn_steam/handshake_messages.h"
#include "wn_steam/key_dictionary.h"
#include "wn_steam/wire_format.h"

namespace wn_steam {

namespace {

constexpr const char* kLogTag = "WnSteamCh";

#define WN_LOGE(...) __android_log_print(ANDROID_LOG_ERROR, kLogTag, __VA_ARGS__)
#define WN_LOGI(...) __android_log_print(ANDROID_LOG_INFO,  kLogTag, __VA_ARGS__)
#define WN_LOGD(...) __android_log_print(ANDROID_LOG_DEBUG, kLogTag, __VA_ARGS__)

// Constant-time byte compare. Used for HMAC tag comparison so a malformed
// envelope can't be distinguished by timing.
bool ct_equal(std::span<const uint8_t> a, std::span<const uint8_t> b) noexcept {
    if (a.size() != b.size()) return false;
    uint8_t diff = 0;
    for (size_t i = 0; i < a.size(); ++i) diff |= a[i] ^ b[i];
    return diff == 0;
}

template <typename Cb, typename... Args>
void safe_invoke(Cb& cb, Args&&... args) {
    if (!cb) return;
    try {
        cb(std::forward<Args>(args)...);
    } catch (const std::exception& e) {
        WN_LOGE("channel callback threw: %s", e.what());
    } catch (...) {
        WN_LOGE("channel callback threw unknown exception");
    }
}

}  // namespace

// ---------------------------------------------------------------------------

EncryptedChannel::EncryptedChannel(std::unique_ptr<Transport> transport)
    : transport_(std::move(transport)) {
    // Wire the transport callbacks to our handlers. The transport's
    // worker thread invokes these — see thread model in the header.
    transport_->set_on_connected([this]() { on_transport_connected(); });
    transport_->set_on_disconnected(
        [this](TransportDisconnectReason r, const std::string& detail) {
            on_transport_disconnected(r, detail);
        });
    transport_->set_on_message(
        [this](std::span<const uint8_t> bytes) { on_transport_message(bytes); });
}

EncryptedChannel::~EncryptedChannel() {
    disconnect();
}

bool EncryptedChannel::connect(const std::string& url) {
    ChannelState expected = ChannelState::Disconnected;
    if (!state_.compare_exchange_strong(expected, ChannelState::Connected)) {
        // The Connected state here means "transport-connecting"; we set it
        // before the actual transport returns so a re-entrant connect call
        // is rejected.
        return false;
    }
    if (!transport_->connect(url)) {
        state_.store(ChannelState::Disconnected);
        return false;
    }
    return true;
}

bool EncryptedChannel::send(std::span<const uint8_t> plaintext) {
    if (state_.load() != ChannelState::Encrypted) return false;
    // WSS path: send plaintext directly. TLS handles encryption at the
    // transport layer; Steam does NOT expect an app-layer AES envelope
    // on WebSocket connections. The encrypt_envelope() helper remains
    // available for a future raw-TCP path.
    return transport_->send(plaintext);
}

void EncryptedChannel::disconnect() {
    ChannelState s = state_.load();
    if (s == ChannelState::Disconnected) return;
    state_.store(ChannelState::Closing);
    transport_->disconnect();
}

void EncryptedChannel::set_on_message(MessageCallback cb) {
    std::lock_guard<std::mutex> lk(cb_mu_);
    on_message_ = std::move(cb);
}
void EncryptedChannel::set_on_connected(ConnectedCallback cb) {
    std::lock_guard<std::mutex> lk(cb_mu_);
    on_connected_ = std::move(cb);
}
void EncryptedChannel::set_on_disconnected(DisconnectedCallback cb) {
    std::lock_guard<std::mutex> lk(cb_mu_);
    on_disconnected_ = std::move(cb);
}

// ---------------------------------------------------------------------------
// Transport callbacks
// ---------------------------------------------------------------------------

void EncryptedChannel::on_transport_connected() {
    // WSS transport is BARE — Steam does NOT do the
    // ChannelEncryptRequest/Response/Result handshake over WebSockets.
    // TLS already encrypts the channel, so there is no app-layer
    // encryption negotiation. Verified against JavaSteam, steam-vent,
    // node-steam-user, and SteamKit2 — all four skip 1303/1304/1305 on
    // WSS and treat the WebSocket as a bare framed transport. The first
    // application message must come from the client (typically
    // CMsgClientHello or CMsgClientLogon).
    //
    // We transition straight to the Encrypted state (which now just
    // means "ready to send/receive Steam-protocol protobufs without an
    // app-layer AES envelope") and fire on_connected_ so CMClient can
    // send the opening frame.
    state_.store(ChannelState::Encrypted);
    WN_LOGI("WSS transport up — skipping ChannelEncrypt handshake (TLS-encrypted bare channel)");

    ConnectedCallback cb;
    {
        std::lock_guard<std::mutex> lk(cb_mu_);
        cb = on_connected_;
    }
    safe_invoke(cb);
}

void EncryptedChannel::on_transport_disconnected(TransportDisconnectReason r,
                                                 const std::string& detail) {
    state_.store(ChannelState::Disconnected);
    DisconnectedCallback cb;
    {
        std::lock_guard<std::mutex> lk(cb_mu_);
        cb = on_disconnected_;
    }
    const auto reason = (r == TransportDisconnectReason::UserInitiated)
        ? ChannelDisconnectReason::UserInitiated
        : ChannelDisconnectReason::TransportError;
    safe_invoke(cb, reason, detail);
}

void EncryptedChannel::on_transport_message(std::span<const uint8_t> bytes) {
    const ChannelState s = state_.load();
    WN_LOGD("rx %zu bytes (state=%d)", bytes.size(), static_cast<int>(s));

    // WSS path: bytes are plaintext Steam-protocol frames (the wire
    // envelope is [u32 emsg|proto_flag][u32 hdr_len][CMsgProtoBufHeader]
    // [body]). Pass through to the upper layer (CMClient) which parses
    // ProtoEnvelope and routes by EMsg.
    //
    // The handshake (ChannelEncryptRequest/Response/Result) is not used
    // on WSS — see on_transport_connected above. The encrypt_envelope /
    // decrypt_envelope helpers remain in this file for a future raw-TCP
    // implementation.
    if (s != ChannelState::Encrypted) {
        WN_LOGE("inbound message in unexpected state %d", static_cast<int>(s));
        return;
    }

    MessageCallback cb;
    {
        std::lock_guard<std::mutex> lk(cb_mu_);
        cb = on_message_;
    }
    safe_invoke(cb, bytes);
}

// ---------------------------------------------------------------------------
// Handshake handlers
// ---------------------------------------------------------------------------

bool EncryptedChannel::handle_channel_encrypt_request(std::span<const uint8_t> body) {
    auto req = MsgChannelEncryptRequest::deserialize_body(body);
    if (!req) return false;
    negotiated_universe_ = req->universe;

    auto spki = get_universe_public_key(req->universe);
    if (spki.empty()) {
        WN_LOGE("no public key for universe=%u", static_cast<uint32_t>(req->universe));
        return false;
    }

    // 1. Generate (or inject for test) the session key.
    SessionKey raw_key;
    if (test_injected_key_.has_value()) {
        raw_key = *test_injected_key_;
    } else {
        auto k = generate_session_key();
        if (!k) {
            WN_LOGE("RNG failure during session-key generation");
            return false;
        }
        raw_key = k->bytes;
    }

    // 2. Assemble plaintext for the RSA envelope: tempSessionKey || challenge.
    std::vector<uint8_t> rsa_input;
    rsa_input.reserve(kSessionKeyLength + req->challenge.size());
    rsa_input.insert(rsa_input.end(), raw_key.begin(), raw_key.end());
    rsa_input.insert(rsa_input.end(), req->challenge.begin(), req->challenge.end());

    auto rsa_ct = rsa_oaep_sha1_encrypt(spki, rsa_input);
    if (!rsa_ct) {
        WN_LOGE("RSA-OAEP encrypt failed");
        return false;
    }
    if (rsa_ct->size() != kRsa1024CipherBytes) {
        WN_LOGE("RSA ciphertext unexpected size=%zu", rsa_ct->size());
        return false;
    }

    // 3. Build and send MsgChannelEncryptResponse.
    MsgChannelEncryptResponse resp;
    std::memcpy(resp.encrypted_handshake_blob.data(), rsa_ct->data(),
                kRsa1024CipherBytes);
    resp.key_crc       = crc32(*rsa_ct);
    resp.unknown_zero  = 0;

    std::vector<uint8_t> out;
    out.reserve(kMsgHdrBytes + kChannelEncryptResponseBodyBytes);
    MsgHdr hdr;
    hdr.msg = EMsg::ChannelEncryptResponse;
    hdr.serialize(out);
    resp.serialize_body(out);

    if (!transport_->send(out)) {
        WN_LOGE("transport->send failed for ChannelEncryptResponse");
        return false;
    }

    // 4. Commit the negotiated key now (we'll use it once Result arrives).
    session_key_  = SecureSessionKey(raw_key);
    std::copy_n(raw_key.begin(), kHmacKeyLength, hmac_key_.begin());

    state_.store(ChannelState::Challenged);
    return true;
}

bool EncryptedChannel::handle_channel_encrypt_result(std::span<const uint8_t> body) {
    auto res = MsgChannelEncryptResult::deserialize_body(body);
    if (!res) return false;
    // EResult.OK == 1 (matches SteamKit's enum).
    if (res->result != 1) {
        WN_LOGE("server EResult for handshake = %u", res->result);
        return false;
    }

    state_.store(ChannelState::Encrypted);
    WN_LOGI("encrypted channel established (universe=%u)",
            static_cast<uint32_t>(negotiated_universe_));

    ConnectedCallback cb;
    {
        std::lock_guard<std::mutex> lk(cb_mu_);
        cb = on_connected_;
    }
    safe_invoke(cb);
    return true;
}

void EncryptedChannel::fail_channel(ChannelDisconnectReason r, std::string detail) {
    state_.store(ChannelState::Closing);
    DisconnectedCallback cb;
    {
        std::lock_guard<std::mutex> lk(cb_mu_);
        cb = on_disconnected_;
    }
    transport_->disconnect();
    safe_invoke(cb, r, std::move(detail));
    state_.store(ChannelState::Disconnected);
}

// ---------------------------------------------------------------------------
// Envelope encrypt / decrypt (NetFilterEncryptionWithHMAC)
//
// Outbound envelope:
//   random_iv      = 13 random bytes
//   hmac_input     = random_iv || plaintext
//   hmac_full      = HMAC-SHA1(hmac_key_, hmac_input)         (20 bytes)
//   iv_plaintext   = hmac_full[0..3] || random_iv             (16 bytes)
//   iv_ciphertext  = AES-256-ECB(session_key, iv_plaintext)
//   body_cipher    = AES-256-CBC(session_key, iv_plaintext, plaintext, PKCS7)
//   on_wire        = iv_ciphertext || body_cipher
//
// Inbound: reverse — verify HMAC in constant time before returning plaintext.
// ---------------------------------------------------------------------------

std::optional<std::vector<uint8_t>>
EncryptedChannel::encrypt_envelope(std::span<const uint8_t> plaintext) const {
    // 13 random bytes
    std::array<uint8_t, kAesBlockBytes - 4> random_iv{};
    if (!secure_random_bytes(random_iv)) return std::nullopt;

    // hmac_input = random_iv || plaintext
    std::vector<uint8_t> hmac_input;
    hmac_input.reserve(random_iv.size() + plaintext.size());
    hmac_input.insert(hmac_input.end(), random_iv.begin(), random_iv.end());
    hmac_input.insert(hmac_input.end(), plaintext.begin(), plaintext.end());

    auto hmac_full = hmac_sha1(hmac_key_, hmac_input);
    if (!hmac_full) return std::nullopt;

    // iv_plaintext = hmac[0..3] || random_iv
    AesBlock iv_plaintext{};
    std::copy_n(hmac_full->begin(), 4, iv_plaintext.begin());
    std::copy_n(random_iv.begin(), random_iv.size(), iv_plaintext.begin() + 4);

    // iv_ciphertext = AES-256-ECB(session_key, iv_plaintext)
    AesBlock iv_ciphertext{};
    if (!aes256_ecb_encrypt_block(session_key_.bytes, iv_plaintext, iv_ciphertext))
        return std::nullopt;

    // body = AES-256-CBC(session_key, iv_plaintext, plaintext, PKCS7)
    auto body = aes256_cbc_encrypt(session_key_.bytes, iv_plaintext, plaintext);
    if (!body) return std::nullopt;

    std::vector<uint8_t> out;
    out.reserve(kAesBlockBytes + body->size());
    out.insert(out.end(), iv_ciphertext.begin(), iv_ciphertext.end());
    out.insert(out.end(), body->begin(), body->end());
    return out;
}

std::optional<std::vector<uint8_t>>
EncryptedChannel::decrypt_envelope(std::span<const uint8_t> wire) const {
    if (wire.size() < kAesBlockBytes + kAesBlockBytes) return std::nullopt;

    AesBlock iv_ciphertext{};
    std::copy_n(wire.begin(), kAesBlockBytes, iv_ciphertext.begin());

    AesBlock iv_plaintext{};
    if (!aes256_ecb_decrypt_block(session_key_.bytes, iv_ciphertext, iv_plaintext))
        return std::nullopt;

    auto body_ct = wire.subspan(kAesBlockBytes);
    auto plaintext = aes256_cbc_decrypt(session_key_.bytes, iv_plaintext, body_ct);
    if (!plaintext) return std::nullopt;

    // Verify HMAC. hmac_input = iv_plaintext[4..16] || plaintext (13 + N).
    std::vector<uint8_t> hmac_input;
    hmac_input.reserve(13 + plaintext->size());
    hmac_input.insert(hmac_input.end(),
                      iv_plaintext.begin() + 4, iv_plaintext.end());
    hmac_input.insert(hmac_input.end(), plaintext->begin(), plaintext->end());

    auto hmac_full = hmac_sha1(hmac_key_, hmac_input);
    if (!hmac_full) return std::nullopt;

    std::span<const uint8_t> hmac_actual(iv_plaintext.data(), 4);
    std::span<const uint8_t> hmac_expect(hmac_full->data(), 4);
    if (!ct_equal(hmac_actual, hmac_expect)) {
        return std::nullopt;  // HMAC mismatch — caller treats as channel kill
    }
    return std::move(*plaintext);
}

}  // namespace wn_steam

File: app/src/main/cpp/wn-steam-client/rust/src/encrypted_channel.rs
use crate::crypto::{
    aes256_cbc_decrypt, aes256_cbc_encrypt, aes256_ecb_decrypt_block, aes256_ecb_encrypt_block,
    hmac_sha1, secure_random_bytes, AesBlock, SecureSessionKey, SessionKey, AES_BLOCK_BYTES,
    HMAC_KEY_LENGTH,
};
use crate::transport::{Transport, TransportDisconnectReason};
use std::sync::atomic::{AtomicU8, Ordering};
use std::sync::{Arc, Mutex};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[repr(u8)]
pub enum ChannelState {
    Disconnected,
    Connected,
    Challenged,
    Encrypted,
    Closing,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[repr(u8)]
pub enum ChannelDisconnectReason {
    UserInitiated,
    TransportError,
    HandshakeProtocolError,
    HandshakeFailed,
    EnvelopeDecryptFailed,
    HmacMismatch,
}

type MessageCallback = Box<dyn Fn(&[u8]) + Send + Sync + 'static>;
type ConnectedCallback = Box<dyn Fn() + Send + Sync + 'static>;
type DisconnectedCallback = Box<dyn Fn(ChannelDisconnectReason, &str) + Send + Sync + 'static>;

pub struct EncryptedChannel {
    inner: Arc<EncryptedChannelInner>,
    transport: Mutex<Box<dyn Transport>>,
}

struct EncryptedChannelInner {
    state: AtomicU8,
    on_message: Mutex<Option<MessageCallback>>,
    on_connected: Mutex<Option<ConnectedCallback>>,
    on_disconnected: Mutex<Option<DisconnectedCallback>>,
}

impl Default for EncryptedChannelInner {
    fn default() -> Self {
        Self {
            state: AtomicU8::new(ChannelState::Disconnected as u8),
            on_message: Mutex::new(None),
            on_connected: Mutex::new(None),
            on_disconnected: Mutex::new(None),
        }
    }
}

impl EncryptedChannel {
    pub fn new(mut transport: Box<dyn Transport>) -> Self {
        let inner = Arc::new(EncryptedChannelInner::default());

        let weak = Arc::downgrade(&inner);
        transport.set_on_connected(Box::new(move || {
            if let Some(inner) = weak.upgrade() {
                inner
                    .state
                    .store(ChannelState::Encrypted as u8, Ordering::Release);
                if let Some(callback) = inner
                    .on_connected
                    .lock()
                    .expect("encrypted channel callback poisoned")
                    .as_ref()
                {
                    callback();
                }
            }
        }));

        let weak = Arc::downgrade(&inner);
        transport.set_on_disconnected(Box::new(move |reason, detail| {
            if let Some(inner) = weak.upgrade() {
                inner
                    .state
                    .store(ChannelState::Disconnected as u8, Ordering::Release);
                let reason = match reason {
                    TransportDisconnectReason::UserInitiated => {
                        ChannelDisconnectReason::UserInitiated
                    }
                    _ => ChannelDisconnectReason::TransportError,
                };
                if let Some(callback) = inner
                    .on_disconnected
                    .lock()
                    .expect("encrypted channel callback poisoned")
                    .as_ref()
                {
                    callback(reason, detail);
                }
            }
        }));

        let weak = Arc::downgrade(&inner);
        transport.set_on_message(Box::new(move |bytes| {
            if let Some(inner) = weak.upgrade() {
                if inner.state() != ChannelState::Encrypted {
                    return;
                }
                if let Some(callback) = inner
                    .on_message
                    .lock()
                    .expect("encrypted channel callback poisoned")
                    .as_ref()
                {
                    callback(bytes);
                }
            }
        }));

        Self {
            inner,
            transport: Mutex::new(transport),
        }
    }

    pub fn connect(&self, url: &str) -> bool {
        if self
            .inner
            .state
            .compare_exchange(
                ChannelState::Disconnected as u8,
                ChannelState::Connected as u8,
                Ordering::AcqRel,
                Ordering::Acquire,
            )
            .is_err()
        {
            return false;
        }
        if !self
            .transport
            .lock()
            .expect("encrypted channel transport poisoned")
            .connect(url)
        {
            self.inner
                .state
                .store(ChannelState::Disconnected as u8, Ordering::Release);
            return false;
        }
        true
    }

    pub fn send(&self, plaintext: &[u8]) -> bool {
        if self.state() != ChannelState::Encrypted {
            return false;
        }
        self.transport
            .lock()
            .expect("encrypted channel transport poisoned")
            .send(plaintext)
    }

    pub fn disconnect(&self) {
        if self.state() == ChannelState::Disconnected {
            return;
        }
        self.inner
            .state
            .store(ChannelState::Closing as u8, Ordering::Release);
        self.transport
            .lock()
            .expect("encrypted channel transport poisoned")
            .disconnect();
    }

    pub fn state(&self) -> ChannelState {
        self.inner.state()
    }

    pub fn set_ca_bundle_path(&self, path: &str) {
        self.transport
            .lock()
            .expect("encrypted channel transport poisoned")
            .set_ca_bundle_path(path);
    }

    pub fn set_on_message<F>(&self, callback: F)
    where
        F: Fn(&[u8]) + Send + Sync + 'static,
    {
        *self
            .inner
            .on_message
            .lock()
            .expect("encrypted channel callback poisoned") = Some(Box::new(callback));
    }

    pub fn set_on_connected<F>(&self, callback: F)
    where
        F: Fn() + Send + Sync + 'static,
    {
        *self
            .inner
            .on_connected
            .lock()
            .expect("encrypted channel callback poisoned") = Some(Box::new(callback));
    }

    pub fn set_on_disconnected<F>(&self, callback: F)
    where
        F: Fn(ChannelDisconnectReason, &str) + Send + Sync + 'static,
    {
        *self
            .inner
            .on_disconnected
            .lock()
            .expect("encrypted channel callback poisoned") = Some(Box::new(callback));
    }
}

impl Drop for EncryptedChannel {
    fn drop(&mut self) {
        self.disconnect();
    }
}

impl EncryptedChannelInner {
    fn state(&self) -> ChannelState {
        match self.state.load(Ordering::Acquire) {
            1 => ChannelState::Connected,
            2 => ChannelState::Challenged,
            3 => ChannelState::Encrypted,
            4 => ChannelState::Closing,
            _ => ChannelState::Disconnected,
        }
    }
}

#[derive(Clone)]
pub struct EncryptedEnvelope {
    session_key: SecureSessionKey,
    hmac_key: [u8; HMAC_KEY_LENGTH],
}

impl EncryptedEnvelope {
    pub fn new(session_key: SessionKey) -> Self {
        let mut hmac_key = [0u8; HMAC_KEY_LENGTH];
        hmac_key.copy_from_slice(&session_key[..HMAC_KEY_LENGTH]);
        Self {
            session_key: SecureSessionKey::new(session_key),
            hmac_key,
        }
    }

    pub fn encrypt(&self, plaintext: &[u8]) -> Option<Vec<u8>> {
        let mut random_iv = [0u8; AES_BLOCK_BYTES - 4];
        if !secure_random_bytes(&mut random_iv) {
            return None;
        }
        self.encrypt_with_random_iv(plaintext, random_iv)
    }

    fn encrypt_with_random_iv(
        &self,
        plaintext: &[u8],
        random_iv: [u8; AES_BLOCK_BYTES - 4],
    ) -> Option<Vec<u8>> {
        let mut hmac_input = Vec::with_capacity(random_iv.len() + plaintext.len());
        hmac_input.extend_from_slice(&random_iv);
        hmac_input.extend_from_slice(plaintext);
        let hmac = hmac_sha1(&self.hmac_key, &hmac_input)?;

        let mut iv_plaintext = [0u8; AES_BLOCK_BYTES];
        iv_plaintext[..4].copy_from_slice(&hmac[..4]);
        iv_plaintext[4..].copy_from_slice(&random_iv);

        let iv_ciphertext = aes256_ecb_encrypt_block(&self.session_key.bytes, &iv_plaintext)?;
        let body = aes256_cbc_encrypt(&self.session_key.bytes, &iv_plaintext, plaintext)?;

        let mut out = Vec::with_capacity(AES_BLOCK_BYTES + body.len());
        out.extend_from_slice(&iv_ciphertext);
        out.extend_from_slice(&body);
        Some(out)
    }

    pub fn decrypt(&self, wire: &[u8]) -> Option<Vec<u8>> {
        if wire.len() < AES_BLOCK_BYTES * 2 {
            return None;
        }
        let iv_ciphertext: AesBlock = wire[..AES_BLOCK_BYTES].try_into().ok()?;
        let iv_plaintext = aes256_ecb_decrypt_block(&self.session_key.bytes, &iv_ciphertext)?;
        let plaintext = aes256_cbc_decrypt(
            &self.session_key.bytes,
            &iv_plaintext,
            &wire[AES_BLOCK_BYTES..],
        )?;

        let mut hmac_input = Vec::with_capacity((AES_BLOCK_BYTES - 4) + plaintext.len());
        hmac_input.extend_from_slice(&iv_plaintext[4..]);
        hmac_input.extend_from_slice(&plaintext);
        let expected = hmac_sha1(&self.hmac_key, &hmac_input)?;
        ct_equal(&iv_plaintext[..4], &expected[..4]).then_some(plaintext)
    }
}

fn ct_equal(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::transport::TransportState;
    use std::sync::atomic::{AtomicBool, Ordering};
    use std::sync::mpsc;

    #[test]
    fn envelope_roundtrips_with_fixed_iv() {
        let key = [7u8; 32];
        let env = EncryptedEnvelope::new(key);
        let encrypted = env
            .encrypt_with_random_iv(b"hello steam", [3u8; AES_BLOCK_BYTES - 4])
            .unwrap();
        assert_ne!(encrypted, b"hello steam");
        assert_eq!(env.decrypt(&encrypted).unwrap(), b"hello steam");
    }

    #[test]
    fn envelope_rejects_hmac_mismatch() {
        let env = EncryptedEnvelope::new([7u8; 32]);
        let mut encrypted = env
            .encrypt_with_random_iv(b"hello steam", [3u8; AES_BLOCK_BYTES - 4])
            .unwrap();
        let last = encrypted.last_mut().unwrap();
        *last ^= 0x55;
        assert!(env.decrypt(&encrypted).is_none());
    }

    #[test]
    fn websocket_channel_skips_app_layer_handshake() {
        let shared = Arc::new(MockTransportState::default());
        let channel = EncryptedChannel::new(Box::new(MockTransport {
            shared: Arc::clone(&shared),
        }));
        let connected = Arc::new(AtomicBool::new(false));
        let connected_cb = Arc::clone(&connected);
        let (tx, rx) = mpsc::channel();
        channel.set_on_connected(move || {
            connected_cb.store(true, Ordering::SeqCst);
        });
        channel.set_on_message(move |bytes| {
            tx.send(bytes.to_vec()).unwrap();
        });

        assert!(channel.connect("wss://cm.example.com:443/cmsocket/"));
        assert_eq!(channel.state(), ChannelState::Connected);
        shared.fire_connected();
        assert_eq!(channel.state(), ChannelState::Encrypted);
        assert!(connected.load(Ordering::SeqCst));

        assert!(channel.send(b"client hello"));
        assert_eq!(shared.sent.lock().unwrap()[0], b"client hello");

        shared.fire_message(b"server frame");
        assert_eq!(rx.recv().unwrap(), b"server frame");
    }

    #[test]
    fn channel_maps_transport_disconnect_reasons() {
        let shared = Arc::new(MockTransportState::default());
        let channel = EncryptedChannel::new(Box::new(MockTransport {
            shared: Arc::clone(&shared),
        }));
        let (tx, rx) = mpsc::channel();
        channel.set_on_disconnected(move |reason, detail| {
            tx.send((reason, detail.to_string())).unwrap();
        });
        assert!(channel.connect("wss://cm.example.com:443/cmsocket/"));
        shared.fire_connected();
        shared.fire_disconnected(TransportDisconnectReason::NetworkError, "net down");
        assert_eq!(
            rx.recv().unwrap(),
            (
                ChannelDisconnectReason::TransportError,
                "net down".to_string()
            )
        );
        assert_eq!(channel.state(), ChannelState::Disconnected);
    }

    #[derive(Default)]
    struct MockTransportState {
        connected: Mutex<Option<crate::transport::ConnectedCallback>>,
        disconnected: Mutex<Option<crate::transport::DisconnectedCallback>>,
        message: Mutex<Option<crate::transport::MessageCallback>>,
        sent: Mutex<Vec<Vec<u8>>>,
        ca_bundle_path: Mutex<String>,
    }

    impl MockTransportState {
        fn fire_connected(&self) {
            if let Some(callback) = self.connected.lock().unwrap().as_ref() {
                callback();
            }
        }

        fn fire_disconnected(&self, reason: TransportDisconnectReason, detail: &str) {
            if let Some(callback) = self.disconnected.lock().unwrap().as_ref() {
                callback(reason, detail);
            }
        }

        fn fire_message(&self, bytes: &[u8]) {
            if let Some(callback) = self.message.lock().unwrap().as_ref() {
                callback(bytes);
            }
        }
    }

    struct MockTransport {
        shared: Arc<MockTransportState>,
    }

    impl Transport for MockTransport {
        fn connect(&mut self, _url: &str) -> bool {
            true
        }

        fn send(&mut self, data: &[u8]) -> bool {
            self.shared.sent.lock().unwrap().push(data.to_vec());
            true
        }

        fn disconnect(&mut self) {}

        fn state(&self) -> TransportState {
            TransportState::Connected
        }

        fn set_on_message(&mut self, cb: Box<dyn Fn(&[u8]) + Send + Sync>) {
            *self.shared.message.lock().unwrap() = Some(cb);
        }

        fn set_on_connected(&mut self, cb: Box<dyn Fn() + Send + Sync>) {
            *self.shared.connected.lock().unwrap() = Some(cb);
        }

        fn set_on_disconnected(
            &mut self,
            cb: Box<dyn Fn(TransportDisconnectReason, &str) + Send + Sync>,
        ) {
            *self.shared.disconnected.lock().unwrap() = Some(cb);
        }

        fn set_ca_bundle_path(&mut self, path: &str) {
            *self.shared.ca_bundle_path.lock().unwrap() = path.to_string();
        }
    }
}
